### PR TITLE
Refactoring YAML tests VersionRange

### PR DESCRIPTION
**Pull request by ldematte**
Split `VersionRange` into an interface + specific implementations; add special "ranges" that do not need Version parsing, and push Version parsing down to were it is really needed (Version range checks).

### CODE REVIEW

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/VersionRange.java [hunk 1]
```diff
@@ -7,42 +7,85 @@
...
+import java.util.Optional;
+import java.util.Set;
 
-public record VersionRange(Version lower, Version upper) {
+public interface VersionRange {
 
-    public boolean contains(Version currentVersion) {
-        return lower != null && upper != null && currentVersion.onOrAfter(lower) && currentVersion.onOrBefore(upper);
-    }
+    boolean matches(Set<String> nodesVersions);
```
Comment (thecoop) [inline test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/VersionRange.java:23]: This definitely could be a `Predicate<Set<String>>`. Either add it as a subinterface, or could use `Predicate<Set<String>>` directly and have a `VersionRanges` helper class?

File: test/yaml-rest-runner/src/test/java/org/elasticsearch/test/rest/yaml/section/VersionRangeTests.java [hunk 2]
```diff
@@ -156,37 +156,37 @@ public void testMatchAll() {
 
         var versionRange = VersionRange.parseVersionRanges(versionRangeString);
 
-        assertTrue(versionRange.get(0).matches(Set.of(randomAlphaOfLength(10))));
+        assertTrue(versionRange.get(0).test(Set.of(randomAlphaOfLength(10))));
```
Comment (thecoop) [inline test/yaml-rest-runner/src/test/java/org/elasticsearch/test/rest/yaml/section/VersionRangeTests.java:159]: See #104062, which will help here

Comment (ldematte) [reply test/yaml-rest-runner/src/test/java/org/elasticsearch/test/rest/yaml/section/VersionRangeTests.java:159]: Nice! Will use it

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/VersionRange.java [hunk 3]
```diff
@@ -7,46 +7,85 @@
...
-    }
+    private VersionRange() {}
+
+    static final Predicate<Set<String>> NEVER = v -> false;
+
+    static final Predicate<Set<String>> ALWAYS = v -> true;
+
+    static final Predicate<Set<String>> CURRENT = versions -> versions.size() == 1 && versions.contains(Build.current().version());
+
+    static final Predicate<Set<String>> NON_CURRENT = versions -> CURRENT.test(versions) == false;
```
Comment (thecoop) [inline test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/VersionRange.java:32]: Could use `CURRENT.not()`

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/VersionRange.java [hunk 4]
```diff
@@ -7,46 +7,85 @@
...
+
+    static final Predicate<Set<String>> MIXED = versions -> versions.size() > 1;
+
+    static class MinimumContainedInVersionRange implements Predicate<Set<String>> {
+        final Version lower;
+        final Version upper;
```
Comment (thecoop) [inline test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/VersionRange.java:38]: Worth documenting somewhere both bounds are inclusive

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/VersionRange.java [hunk 5]
```diff
@@ -7,46 +7,85 @@
...
-    @Override
-    public String toString() {
-        return "[" + lower + " - " + upper + "]";
+        @Override
+        public boolean test(Set<String> nodesVersions) {
+            // Try to extract the minimum node version
+            var minimumNodeVersion = nodesVersions.stream()
+                .map(ESRestTestCase::parseLegacyVersion)
+                .flatMap(Optional::stream)
+                .min(VersionId::compareTo)
```
[the remark this answers is not in the document]
Comment (thecoop) [reply test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/VersionRange.java:51]: Oh yes, quite right...

**Review by thecoop — APPROVED** [review]

File: test/yaml-rest-runner/src/test/java/org/elasticsearch/test/rest/yaml/section/VersionRangeTests.java [hunk 6]
```diff
@@ -70,4 +87,108 @@ public void testParseWithThreeDigitVersion() {
...
+        String versionRangeString = "6.0.0 - 6.1.0, 7.1.0 - 7.5.0";
+
+        var versionRanges = VersionRange.parseVersionRanges(versionRangeString);
+
+        var match1 = versionRanges.stream().filter(range -> range.test(Set.of("6.1.0"))).findFirst();
+        var match2 = versionRanges.stream().filter(range -> range.test(Set.of("7.1.0"))).findFirst();
+        var nonMatch1 = versionRanges.stream().filter(range -> range.test(Set.of("5.1.0"))).findFirst();
+        var nonMatch2 = versionRanges.stream().filter(range -> range.test(Set.of("8.1.0"))).findFirst();
+
+        assertTrue(match1.isPresent());
```
Comment (thecoop) [inline test/yaml-rest-runner/src/test/java/org/elasticsearch/test/rest/yaml/section/VersionRangeTests.java:128]: See #104123, which could help here

**Comment by ldematte** [comment]
@elasticmachine run elasticsearch-ci

**Comment by ldematte** [comment]
@elasticmachine retest this please

**Comment by ldematte** [comment]
@elasticmachine update branch

**Pull request merged**